sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -356,25 +356,24 @@ case class CreateNamedStruct(children: Seq[Expression]) extends CreateNamedStruc
   override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = {
     val rowClass = classOf[GenericInternalRow].getName
     val values = ctx.freshName("values")
-    ctx.addMutableState("Object[]", values, s"$values = null;")
+    ctx.addMutableState("Object[]", values, s"$values = new Object[${valExprs.size}];")
     val valuesCode = ctx.splitExpressionsWithCurrentInputs(
       valExprs.zipWithIndex.map { case (e, i) =>
         val eval = e.genCode(ctx)
         s"""
-          ${eval.code}
-          if (${eval.isNull}) {
-            $values[$i] = null;
-          } else {
-            $values[$i] = ${eval.value};
-          }"""
+           |${eval.code}
+           |if (${eval.isNull}) {
+           |  $values[$i] = null;
+           |} else {
+           |  $values[$i] = ${eval.value};
+           |}
+         """.stripMargin
       })
 
     ev.copy(code =
       s"""
-         |$values = new Object[${valExprs.size}];
          |$valuesCode
          |final InternalRow ${ev.value} = new $rowClass($values);
-         |$values = null;
```

Comment:
We create the object array every time is because `GenericInternalRow` doesn't copy the given array. If the parent operator keeps the produced rows, this change may cause wrong result.

Reply:
I had the same feeling. But then I thought: isn't the same for each other value?

Reply:
I think the contractor is, if we need to buffer some values, we need to copy them. BTW `CreateArray` also reuse the array.

Reply:
Actually `CreateArray` only reuse the array if element type is not primitive, we should also fix it in another PR. @mgaido91

Reply:
IIUC I think `CreateArray` never reuses the array. Why are you saying it does?

Reply:
damn it was reverted in https://github.com/apache/spark/pull/19797/files#diff-c1758d627a06084e577be0d33d47f44eL97 and I was reading the old code. Let's bring it back.

Reply:
also cc @kiszk , is it by accident?

Reply:
#19797 addressed to avoid using an object in a global variable in several places since to reuse an object looks [a hacky way](https://github.com/apache/spark/pull/19797#discussion_r152766978).  
I intentionally avoided the reuse of an array.

Reply:
sorry I didn't remember it very clearly, what was the gain of not reusing the array? Saving a global variable slot seems not worth.